lib/mutable_tree_map.fz

```diff
@@ -0,0 +1,320 @@
+    # node, or -1 if an empty subtree is given
+    #
+    private height(node option (Entry LM K V)) i32 is
+      match node
+        nil => -1
+        e Entry => e.height.get
+
+
+    # returns the (AVL) balance factor of the given node, or
+    # 0 if a nil node is given
+    #
+    private balance_factor(node option (Entry LM K V)) i32 is
+      match node
+        nil => 0
+        e Entry => (height e.right.get) - (height e.left.get)
+
+
+    # recalculates and updates the heights of the subtrees in the
+    # subtree whose root is the given node
+    #
+    private fix_height(node option (Entry LM K V)) =>
+      match node
+        nil =>
+        e Entry =>
+          lh := height e.left.get
+          rh := height e.right.get
+
+          e.height <- (if lh > rh then lh else rh) + 1
+
+
+    # rotate right at the given node
+    #
+    private rotate_right(node option (Entry LM K V)) option (Entry LM K V) is
+      # because this feature is only called when the tree is out of balance,
+      # i.e. the left subtree has more nodes than the right one, we can safely
+      # assume here that node and node.left are not nil.
+      l := node.get.left.get
+
+      node.get.left <- l.get.right.get
+      l.get.right <- node
+
+      fix_height node
+      fix_height l
+
+      l
+
+
+    # rotate left at the given node
+    #
+    private rotate_left(node option (Entry LM K V)) option (Entry LM K V) is
+      # because this feature is only called when the tree is out of balance,
+      # i.e. the right subtree has more nodes than the left one, we can safely
+      # assume here that node and node.right are not nil.
+      r := node.get.right.get
+
+      node.get.right <- r.get.left.get
+      r.get.left <- node
+
+      fix_height node
+      fix_height r
+
+      r
+
+
+    # rebalance is called when nodes might have been inserted or deleted.
+    # this means that the heights of the nodes potentially changed, thus
+    # we need to recalculate them.
+    fix_height node
+
+    bf := balance_factor node
+
+    if bf < -1
+      # we can safely do node.get here because if node was empty,
+      # its balance factor would be 0.
+      if balance_factor node.get.left.get <= 0
+        rotate_right node
+      else
+        node.get.left <- rotate_left node.get.left.get
+        rotate_right node
+    else if bf > 1
+      # we can safely do node.get here because if node was empty,
+      # its balance factor would be 0.
+      if balance_factor node.get.right.get >= 0
+        rotate_left node
+      else
+        node.get.right <- rotate_right node.get.right.get
+        rotate_left node
+    else
+      # do nothing
+      node
+
+
+  # returns an empty tree of elements of type A.
+  #
+  type.empty mutable_tree_map LM K V is
+    mutable_tree_map LM K V
+
+
+  # returns a tree of elements of type A that contains just the element a.
+  #
+  type.singleton(k K, v V) mutable_tree_map LM K V is
```

Comment:
`type.new` would be handy, similar to `map_of(K type : has_total_order, V type, kvs array (tuple K V))`